records/2021-07-16.scalability.database-bottleneck.md

```diff
@@ -0,0 +1,135 @@
+# Database performance issues
+
+Date: 2021-07-16
+
+## Status
+
+**DRAFT**
+
+## Context
+
+Because Notify suffers from regular daily incident that originates from its
+database when clients blasts notification, we need to review some design aspect
+of the stack and bring more control to specific elements.
+
+In the past months, the current architecture, environment and configuration have
+been tweaked to achieve a certain balance that could take on the expected load.
+To list a few of these tweaks:
+
+* the number of maximum pool connection in individual application instance,
+* the AWS instance classes,
+* the resources requests and limits of Kubernetes containers,
+* the horizontal pod autoscaling.
+
+These serve well the application but there were already signs of the
+application's limit: we rarely if ever meet our SLAs and we had regular daily
+incidents that foretold of the difficulties ahead.
+
+This document propose various options to address thorny issues that our
+database is having.
+
+## Timeline of an incident
+
+1. A blast of notifications is sent to Notify through the API.
+2. As the notification-api component saves the notifications to be sent in
+the database, it commits each of these individually.
+3. The notification is then sent to the appropriate queue, depending on its
+type, scheduling and used key.
+4. In the case of an email that was requested through the API, the email
+will be sent to the AWS SES provider (the only supported at the time of
+this writing).
+5. As multiples emails are sent and some receive a response receipt, multiple
+updates on the database are performed.
+6. In the meantime, numerous count operations are performed to track the
+current delivered emails and various metrics.
+7. As there are more and more commits performed on the database, for each and
+every notification, the database triggers an autovacuum action to clean up the
```

Comment:
does the autovacuum slow dow the whole database or just activities on one table?

Reply:
It slows down all components interacting with the database, so I infer the whole database for now.